sdk/metric/sdk.go

```diff
@@ -229,32 +241,50 @@ func (m *SDK) SetErrorHandler(f ErrorHandler) {
 	m.errorHandler = f
 }
 
-func (s *syncInstrument) acquireHandle(ls *labels) *record {
-	// Create lookup key for sync.Map (one allocation)
+// acquireHandle gets or creates a `*record` corresponding to `kvs`,
+// the input labels.  The second argument `labels` is passed in to
+// support re-use of the orderedLabels computed by a previous
+// measurement in the same batch.   This performs two allocations
+// in the common case.
+func (s *syncInstrument) acquireHandle(kvs []core.KeyValue, lptr *labels) *record {
+	var rec *record
+	var labels labels
+
+	if lptr == nil || lptr.ordered == nil {
+		// This memory allocation may not be used, but it's
+		// needed for the `sortSlice` field, to avoid an
+		// allocation while sorting.
+		rec = &record{}
+		labels = s.meter.makeLabels(kvs, &rec.sortSlice)
+	} else {
+		labels = *lptr
+	}
+
+	// Create lookup key for sync.Map (one allocation, as this
+	// passes through an interface{})
 	mk := mapkey{
 		descriptor: &s.descriptor,
-		ordered:    ls.ordered,
+		ordered:    labels.ordered,
 	}
 
 	if actual, ok := s.meter.current.Load(mk); ok {
-		// Existing record case, only one allocation so far.
-		rec := actual.(*record)
-		if rec.refMapped.ref() {
+		// Existing record case.
+		newRec := actual.(*record)
```

Comment:
Could the name be rather `existingRec`?

And now I can see the bug too. It was the (unwanted?) reuse of the unmapped record, which resulted in a kind of a race between removing it from the map and loading it from the map in a loop below. The reuse caused the reference count to be reset from (0, unmapped) to (2, mapped), then we could load it from the map, `oldRec.refMapped.ref()` would return true (because it's the reused record), so we return the reused record. Context switches back to the collect goroutine that removes the reused record from the map. Boom.

Makes me wonder… Could the reuse work if we had the code like this:

```go
if actual, ok := s.meter.current.Load(mk); ok {
	rec = actual.(*record)
	if rec.refMapped.ref() {
		// At this moment it is guaranteed that the entry is in
		// the map and will not be removed.
		return rec
	}
	// This entry is no longer mapped, try to add a new entry.
	// New stuff here:
	// Try to reuse the entry anyway
	for {
		// let the collect goroutine remove the unmapped record from the map
		runtime.Gosched()
		oldRec, ok := s.meter.current.Load(mk)
		// reset is a new function that would reset ref count to 2 and with mapped state atomically
		// if the previous refmapped was at ref count at 0 and unmapped state
		if !ok {
			// entry is gone from the map
			if !rec.refmapped.reset() {
				// some other goroutine reused `rec` already
				rec = nil
				// TODO: this probably means that other goroutine took rec,
				// but haven't yet managed to put it into the map,
				// do not set rec to nil, but just continue?
			}
			break
		}
		if rec != oldRec {
			// some other goroutine already reused rec,
			// but maybe oldRec is also already unmapped
			if oldRec.refMapped.reset() {
				// oldRec is ours
				rec = oldRec
			} else {
				// nothing to reuse
				rec = nil
				// TODO: ref oldRec and return oldRec?
			}
			break
		} else if rec.refcount.mapped() {
			// rec got reused
			rec = nil
			// TODO: ref rec and return ref?
		}
	}
}
```

Sorry, got carried away a bit, but was fun to write. :P Anyway, is the reuse of `rec` a fringe case? (Outside stress testing, that is.)

Reply:
I renamed to `existingRec`. Yes, the bug was introduced when I made the optimization to `RecordBatch`. 😁 

I wasn't thinking about whether we could re-use the record. The idea scares me a bit, probably a fringe case. I added a related TODO though, about how we might consider leaving records in the map for the next collection interval. That was the behavior prior to #468.